diff --git a/lib/Driver/Tools.cpp b/lib/Driver/Tools.cpp
index 01d6844..658e7ac 100644
--- a/lib/Driver/Tools.cpp
+++ b/lib/Driver/Tools.cpp
@@ -5926,8 +5926,14 @@ void darwin::Link::ConstructJob(Compilation &C, const JobAction &JA,
   Args.AddAllArgs(CmdArgs, options::OPT_L);
 
   if (Args.hasArg(options::OPT_fopenmp)){
+    const toolchains::Linux& ToolChain = static_cast<const toolchains::Linux&>(getToolChain());
+    const SanitizerArgs &Sanitize = ToolChain.getSanitizerArgs();
+
     // This is more complicated in gcc...
-    CmdArgs.push_back("-liomp5");
+    if(Sanitize.needsTsanRt())
+      CmdArgs.push_back("-liomp5_tsan");
+    else
+      CmdArgs.push_back("-liomp5");
 
     if (Args.hasArg(options::OPT_omptargets_EQ))
       CmdArgs.push_back("-lomptarget");
@@ -7668,12 +7674,18 @@ void gnutools::Link::ConstructJob(Compilation &C, const JobAction &JA,
 
   if (!Args.hasArg(options::OPT_nostdlib)) {
     if (!Args.hasArg(options::OPT_nodefaultlibs)) {
+      const toolchains::Linux& ToolChain = static_cast<const toolchains::Linux&>(getToolChain());s
+      const SanitizerArgs &Sanitize = ToolChain.getSanitizerArgs();
       if (Args.hasArg(options::OPT_static))
         CmdArgs.push_back("--start-group");
 
       bool OpenMP = Args.hasArg(options::OPT_fopenmp);
       if (OpenMP) {
-        CmdArgs.push_back("-liomp5");
+	if(Sanitize.needsTsanRt())
+	  CmdArgs.push_back("-liomp5_tsan");
+	else
+	  CmdArgs.push_back("-liomp5");
+
         if (Args.hasArg(options::OPT_omptargets_EQ))
           CmdArgs.push_back("-lomptarget");
       }
